dashboard refactor pt4

mlippert opened the pull request:
## Description
### Moved all Metrics components into a `Metrics` subfolder of Dashboard
`ChartCard`, `TurnChart`, `InfluenceChart`, `TimelineChart`, `gantt`

### Extract child components (existing or new) into their own source files
All use the the new common style for components.
`ChartCardInfo`, `MeetingList`, `MeetingListElement`, `WavyBackground`

### styled components
moved styled components into a single `styled.js` file in the same directory as the component(s) that use them.

### Moved relevant redux state connections from the Dashboard to each Metric component
The `index.js` file in the Metrics folder does all of the redux state connecting for all components in that folder, and exports all of the connected components for use by the Dashboard component.

This also means that each metric handles it's data being loaded asynchronously and displays a loading view, until it is available.

This allowed updating the `loadMeetingData` dashboard action creator, to now update the status of each piece of data as it is loaded, instead of waiting for ALL of them to complete.

### TimelineChart / gantt
- Intensive rewrite of the d3 gantt chart class, providing `draw`, `erase` and `redraw` methods (redraw is currently a poor man's implementation, ie it erases and then draws again, but it seems fine for now).
- The revamped gantt class, fixed several issues w/ the TimelineChart (some of which previous refactor may have introduced). Such as clicking on the info button and then closing it added another chart below the existing chart(s). It also allowed code in the TimelineChart to handle a window resize so that the gantt chart could be redrawn to fit in the new size.

### ChartCard
- Replace the 2 specialized properties (`chartDiv` and `chartTable`) by just rendering all children nodes.

### ChartTable
This component is really for accessibility (a11y) so it was moved out of Dashboard and into it's own `A11y` components folder

### colors
More work to use color names defined in libs/utils `Colors`, so it is easier to see and find all the places a color is used, and to know that the color being used is the same as a color used somewhere else.

### fix style issues
A few files were touched to use Color names or just came into my view, and therefore I took the opportunity to fix eslint reported style issues in those files. `App.js`, `Home.js`, `LeaveRoomButton.jsx`, `MakeMeetingCardView.js`, `MakeMeetingCard/index.js`

### Other
- small modifications to eslint config
- added eslint config override for the Dashboard component files (differences like indentation for when they will be extracted into a new repo/package)
- fixed a bulma issue that was being reported when building

### Work that still should be done
- Extract the MeetingInfo that is displayed in the Dashboard to its own component (probably in Metrics). Hook it up to the redux data, and status it needs, so those can be removed from the Dashboard.
- The TimelineChart (and gantt chart) use the task name (namely the user's name) as the link to the task data (utterances) displayed. This causes an issue when multiple participants have the same name. We should use the participant Id as the link, while still displaying the name.
- The TimelineChart size really should be easier to configure w/ props, right now we didn't set it, instead just relying on hard-coded computed values. I'm not quite sure how to do that, ie should the size be just about the gantt chart, or about the entire TimelineChart, ChartCard and all?

## Motivation and context
This is part of the work started in _Dashboard: Refactoring for modularity_ #83 and being continued in _Dashboard: Continue Refactoring for modularity_ #112

## How has this been tested?
Tested locally on my developer machine using `make up` from riff-docker.

## Types of changes
<!--- What types of changes does your code introduce? -->
<!--- Please remove all lines which don't apply. -->
- ✅ Code refactor (non-breaking change)

## Checklist:
<!--- Please go over all the following points. -->
<!--- Again, remove any lines which don't apply. -->
<!--- Pull Requests that don't fulfill all [REQUIRED] requisites are likely -->
<!--- to be sent back to you for correction or will be rejected.  -->
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**

jordanreedie commented on src/components/Dashboard/Metrics/TimelineChart.jsx:
```diff
@@ -0,0 +1,270 @@
+
+        /** meeting whose timeline data will be in timelineData */
+        meeting: PropTypes.shape({
+            _id: PropTypes.string.isRequired,
```
I'm not super clear on exactly how PropTypes.shape works (and a quick google has not elucidated this for me), does this mean that the meeting object that's being passed *requires* an `_id` property, or that the only property the meeting object can have is `_id`? If the latter, why the change from string to object?

mlippert replied:
I have to say the prop-types documentation is really lacking. I believe it's "he meeting object that's being passed requires an _id property". I'm pretty sure it doesn't care about any other properties of the meeting object. Now if you reference one of them I'd hope that you'd get a warning that it wasn't defined in propTypes, but I don't know if you would.
The change from just passing the meeting id to passing the meeting object is because what is selected (and in the redux state) is the meeting object. And perhaps there will be other properties of a meeting object it would be desirable to reference (I realize that is a very slippery slope, but I am sort of anticipating that the meeting object will provide its participants at some point)

jordanreedie replied:
Ok, that's fair enough!

jordanreedie commented on src/components/Dashboard/Metrics/TimelineChart.jsx:
```diff
@@ -0,0 +1,270 @@
+            .reduce((pmap, p, i) => {
+                pmap[p.id] = {
+                    name: p.name,
+                    color: p.id === participantId ? PeerColors[0] : PeerColors[(i % (PeerColors.length - 1)) + 1],    // eslint-disable-line no-extra-parens
```
for the condition here - wasn't `participantId` filtered out already?

mlippert replied:
I didn't revisit this code in this latest refactor, but yes you're right (that check is an unnecessary hold over and should be removed.

jordanreedie commented on src/components/Dashboard/Metrics/TimelineChart.jsx:
```diff
@@ -0,0 +1,270 @@
+                return pmap;
+            }, participantMap);
+
+        const numOtherPeers = sortedParticipants.length - (participantMap[participantId] ? 1 : 0);
```
when would the participantMap for a meeting the user participated in not include the current user? Is `sortedParticipants` derived from utterances and this is for the case where the user doesn't speak at all in a meeting?

mlippert replied:
Yes, I'm pretty sure that is what's going on. Passing better data is still on the to do list.

mlippert commented on src/components/Dashboard/Metrics/WavyBackground.jsx:
```diff
@@ -0,0 +1,69 @@
+
+import { StyledWave } from './styled';
+
+/* ******************************************************************************
```
The file comment and the class comment are 2 different things with different purposes.
It is very true, particularly in a file whose purpose is defining a single class that the description for the file and the class are similar (or even identical). That doesn't mean (to me) that you should delete one of them.

I believe every file should have a file comment with a short summary, possibly more exposition about the file, including the authors, the creation date and the copyright notice.

And every class should also have a class header comment describing its purpose. It can be tough to say a whole lot, and particularly for small classes there just isn't much to say, but the header should still exist.

It provides the easiest source to understand the purpose of the class when all you want is an overview w/o digging into the actual implementation. (That pretty much goes for function/method comments also). Function method block comments (for me) also make it easy to scroll through a source file and see all the functions. When there are no block comments it's easy to miss them. That's why my style looks like it does.

jordanreedie replied:
Ah, ok. I'm personally not a super big fan of having a class comment when we have the file comment if we're sticking to one component per file. IMO, unless there is additional information that needs to be conveyed, it just adds an extra thing to maintain and potentially get out of sync & cause confusion. That being said, I understand why you want them and I don't have particularly strong feelings about it so I'll concede this one

mlippert replied:
Thanks :smile:

mlippert commented on src/components/Dashboard/Metrics/TimelineChart.jsx:
```diff
@@ -0,0 +1,270 @@
+            //  - the size of the chart should be specified in a prop somehow, not sure
+            //    if the size should be for the whole chart, chart card and all or just the
+            //    chartDiv. -mjl 2019-05-31
+            return (
```
That makes sense, but I like the explicitness of 2 return paths in this situation, because I think of the loading path as "exceptional" and the path that displays the chart as the normal case, and I wanted that to be obvious.

I have been thinking about how to name a method to return the loading jsx, in order to make that "exception" case smaller in the render, I just hadn't gotten there yet. And using a single var like `content` may be the way to go also.

There is definitely more "refactor" to be done to get the jsx looking as clean as you've got in chat (particularly in the dashboard itself).

jordanreedie replied:
Oh ok, I wasn't sure if this was the last refactor pass or not, I don't think you need to address it immediately.

mlippert replied:
I was just looking at it again, And the same way it was bugging you, it's been bugging me. I prefer the 2 return paths, but I'd like a method that takes the "content" and returns in wrapped in that common stuff. I just haven't figure out how to name that method yet.

jordanreedie replied:
Haha yeah just looking at it, my immediate expectation is the two return paths must be totally different, and I think it's the subversion of that expectation that's annoying. I can't think of a good name for it either, though

jordanreedie commented on src/components/Dashboard/Metrics/TimelineChart.jsx:
```diff
@@ -0,0 +1,270 @@
+import { ChartCard } from './ChartCard';
+import { WavyBackground } from './WavyBackground';
+
+const chartInfo = 'This chart shows a timeline of when people spoke during your meeting.';
```
oh also why is `chartInfo` declared up here but `chartTitle` declared in the `render` function? It looks like they're both used in the same places

mlippert replied:
That is a very good question. And they should be consistent.

I think it is more efficient to declare private constants at the module level, rather than in a function. (I'm not sure about that).

I probably was trying to see how I felt about the module level constants, and ended up doing things inconsistently. Which way do you think they should go? I'm currently leaning towards moving the `chartCardTitle` to the module level next to `chartInfo`.

jordanreedie replied:
Hmmm I could probably be swayed either way, I think. When you say 'more efficient', do you mean for performance or human-readability? I can't really imagine it having much of an effect on actual performance. 

The reason I noticed that these were in different places is because when I was looking over the render code, I realized I had no idea where `chartInfo` was coming from. So, I think for readability's sake, it might be better to have it in the render function? I think that's consistent with most of our other code also.

mlippert replied:
I did mean performance, and yes in this case it is totally something that would be negligible, I just hate the idea that something that only needs to be created once, might be created every time a function is run.

The TurnChart actually defines both constants at the module level.

mlippert replied:
OK after some discussion w/ @jordanreedie in slack, the info and title properties are now in a `chartConfig` module level object in both TurnChart and TimelineChart.

mlippert commented:
OK, I've committed updates so that the TurnChart and TimelineChart now define a module level `chartConfig` with cardTitle and info properties.
And I've removed that redundant check of the participant ID when assigning other's colors in the timeline.
And lastly I added a `_decorateContent` method in both the TurnChart and TimelineChart so that the decoration elements aren't duplicated in the `render` implementation.

jaedoucette commented on src/components/Dashboard/Metrics/gantt.js:
```diff
@@ -28,152 +39,166 @@ const TimeDomainMode = {
-        this.privateProp.width = Number(w);
+        let newWidth = w;
+        if (!newWidth) {
+            newWidth = document.body.clientWidth / 2;
```
Same expr.

mlippert replied:
however, yes this is a possibly a case of resetting an initial value.

Actually though I think this is somewhat different, and the 2 expressions while the same are not really serving the same purpose. This case is more error handling, w/ a log message to explicitly call that out. The fact that I chose to use the same expression as is currently used to initialize the backing value is happenstance  (well not really, but I could see changing what the initial value is set to and not changing this).
So I'm leaving this as is.

jaedoucette commented on src/components/Dashboard/Metrics/gantt.js:
```diff
@@ -387,7 +416,7 @@ function wrapYLabels(yLabels, width) {
         let numLines = 1;
         let word;
-        while (word = words.pop()) {
+        while (word = words.pop()) {    // eslint-disable-line no-cond-assign
```
Consider changing this to a for loop to avoid the eslint and to make more explicit?

mlippert replied:
I thought about it, and I'll probably think about it again, but I decided to leave this for now as matching the original source where this function was copied from.

jaedoucette commented:
> These are comments Brec and I's joint review.

@brecriffs

mlippert commented on src/components/Dashboard/ChartCard.jsx:
```diff
@@ -14,15 +14,10 @@
 
-// Set eslint indentation to 4 instead of 2 in preparation to moving this to a new
-// repository, where that will be the rule.
 // I haven't figured out for sure if we(Riff) want to use the underscore prefix to
```
I'm not ready to do that, because I'm still on the fence about the style. If we decide that we want to use the '_' prefix to denote class properties that are considered private, then I'd move that rule to the eslint config for the src tree (and probably also to the config for the server tree). Having the override in particular files makes it easy to find where it being used, and fix it if we decide against it.

jaedoucette replied:
That makes sense. I did some more reading, and it looks like as of the end of March, the JS standards support `#` as an interpreter-enforced privacy prefix. Apparently Chrome's interpreter already ships with this feature, and presumably others will follow. 

In light of that, maybe we should migrate to the `#`-prefix as well?

See: https://github.com/tc39/proposal-class-fields/blob/master/PRIVATE_SYNTAX_FAQ.md

and 

https://stackoverflow.com/questions/4484424/underscore-prefix-for-property-and-method-names-in-javascript

jaedoucette replied:
That makes sense. I did some more reading, and it looks like as of the end of March, the JS standards support `#` as an interpreter-enforced privacy prefix. Apparently Chrome's interpreter already ships with this feature, and presumably others will follow. 

In light of that, maybe we should migrate to the `#`-prefix as well?

See: https://github.com/tc39/proposal-class-fields/blob/master/PRIVATE_SYNTAX_FAQ.md

and the accepted answer to:

https://stackoverflow.com/questions/4484424/underscore-prefix-for-property-and-method-names-in-javascript

jaedoucette replied:
That makes sense. I did some more reading, and it looks like as of the end of March, the JS standards support `#` as an interpreter-enforced privacy prefix. Apparently Chrome's interpreter already ships with this feature, and presumably others will follow. 

In light of that, maybe we should migrate to the `#`-prefix as well?

See: https://github.com/tc39/proposal-class-fields/blob/master/PRIVATE_SYNTAX_FAQ.md

and the accepted answer to:

https://stackoverflow.com/questions/4484424/underscore-prefix-for-property-and-method-names-in-javascript

mlippert replied:
When ecmascript officially supports a method of specifying private fields we should definitely use that. We can't use the `#` prefix now, as it isn't a valid character for a variable name.

mlippert commented on src/components/Dashboard/DashboardView.jsx:
```diff
@@ -74,9 +73,53 @@ class MeetingView extends React.Component {
         handleMeetingClick: PropTypes.func.isRequired,
     }
 
+    /**
```
It's not general purpose enough in its current form to be in utils, it really does something very specific for this class.

jaedoucette replied:
It's not clear from the code snippet attached to this comment, but I think we meant the data/time functionality that got pulled out of the moment package. Looking again, it seems pretty general, and also pretty unrelated to the class (apart from the fact that only this class wants that functionality right now?). 

Ultimately fine if it is specific to this class, but wanted to make sure we were talking about the same thing.

mlippert replied:
Well this comment is on code that isn't in the final commit in this file, but is in the `MeetingListElement.jsx` file where there is a static method `_formatMeetingTime`.
I don't consider formatting a date/time to a very specific string format, to be general purpose, and yes this is the only place that format of string is used. Some of the other functions are almost general purpose enough to warrant being extracted to the utils lib, but I don't think they're quite ready, I'd want to see them needed somewhere else first.

mlippert commented on src/components/Dashboard/index.js:
```diff
@@ -61,15 +60,18 @@ const mapDispatchToProps = dispatch => ({
     const endDate = meeting.endTime ? new Date(meeting.endTime) : new Date();
-    const diff = moment(endDate).diff(moment(startDate), 'minutes');
-    return `${diff} minutes`;
+    const durationMins = Math.trunc((endDate.getTime() - startDate.getTime()) / MILLISEC_PER_MINUTE);
```
Hmm, it doesn't strike me as particularly less readable, and it some ways I find the new code easier to understand what is being done, so I think we'll just disagree on this one.

mlippert commented on src/components/Dashboard/index.js:
```diff
@@ -72,11 +71,10 @@ const formatMeetingDuration = (meeting) => {
-        meeting = { ...meeting, endTime: new Date() };
-    }
-    const diff = moment(new Date(meeting.endTime)).diff(moment(new Date(meeting.startTime)), 'minutes');
+    // support showing data on 'in progress' meeting by giving it a fake end time of now
```
good point. I'll update the comment.

mlippert commented on src/components/Dashboard/.editorconfig:
```diff
@@ -0,0 +1,19 @@
+# http://editorconfig.org
```
It's completely different from an eslint configuration file, did you read the website?
It's a hint file for editors if they support it to set things like use tabs or spaces for various files in the hierarchy. It doesn't have anything to do with how you actually end up structuring your code.

There is an .editorconfig file already at the top level, I added this one here for the same reason as I added the override eslint config, I want to use different indentation for the files in Dashboard, and I was tired of my editor setting the indentation differently than what I wanted to use.

mlippert commented on src/.eslintrc.yml:
```diff
@@ -179,6 +179,8 @@ rules:
     - error
     -
       functions: false
+      classes: false
```
In general you shouldn't be executing anything in a module, therefore a forward reference within a function to something defined later in the module should be safe, and is desirable occasionally, more often than the message not to do that is helpful in my opinion.

mlippert commented on src/components/Dashboard/MeetingList.jsx:
```diff
@@ -0,0 +1,108 @@
+     * _getMeetingTiles (private)                                          */ /**
+     *
+     * Returns an array of MeetingListElements for all of the meetings that
+     * belong in this MeetingList.
```
Yes all meetings supplied in the prop belong in the list. How would you word the method description?

mlippert replied:
That's very specific. As I think about it more I thing _belong_ may be the right term after all as that is the purpose of the method, and what is expected where it is being used. The fact that **everything** currently belongs is an implementation detail that I think it is reasonable to read the code if you want to know that. Well maybe what belongs should also be in the description.

OK I've reworked the description and pushed it.

mlippert commented on src/components/Dashboard/MeetingList.jsx:
```diff
@@ -0,0 +1,108 @@
+     */
+    _getMeetingTiles() {
+        return this.props.meetings
+            .map((meeting) => {
```
really? `selected` isn't clearly a boolean indicating whether the current meeting is the selectedMeeting?

mlippert commented on src/components/Dashboard/MeetingListElement.jsx:
```diff
@@ -0,0 +1,166 @@
+     *
+     * @returns {string} a formatted representation of the given date
+     */
+    static _formatMeetingTime(dt) {
```
see earlier discussion

mlippert commented on src/components/Dashboard/MeetingListElement.jsx:
```diff
@@ -0,0 +1,166 @@
+}
+
+/** Array of english month abbreviations */
+const monthAbbrs = [ 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec' ];
```
All of this is very specific to creating a particular string for a particular purpose. Certainly the moment formatter is more succint, and potentially clearer, but I really wanted to try to remove the couple of small uses we had for that library which is very large.

mlippert commented on src/components/Dashboard/DashboardView.jsx:
```diff
@@ -133,7 +152,8 @@ class DashboardView extends React.Component {
         }
-        else if (fetchMeetingsStatus === 'error') {
+
+        if (fetchMeetingsStatus === 'error') {
```
I think it's clearer that previous condition was handled there is no path that continues, which isn't true if you still have the else.
I like the style of handle simple cases/error conditions and  return up front with the more normal "interesting" functionality following that.
The mattermost team had turned on the lint to flag else after an if that returned, and I agreed with that as a general practice.
I could see converting this particular set of tests to it's own table lookup possibly in a future refactor, although with only 3 cases I'm not sure it would be worth it.

mlippert commented on src/components/Dashboard/DashboardView.jsx:
```diff
@@ -145,12 +165,86 @@ class DashboardView extends React.Component {
                                                               : selectedMeeting.room;                                   // eslint-disable-line indent
 
+        let meetingInfo;
+        if (!selectedMeeting || statsStatus !== 'loaded') {
```
It shouldn't be possible to have loaded stats if a meeting isn't selected, because I'm not sure what that would even mean. But the way that some of these properties are updated in the redux store I felt it's better to be safe if that occurs rather than access an undefined variable. How stuff is represented and updated in the store is something for a future refactor.

Note that this is only for the meeting info, to be displayed or not displayed. And, this is actually one of those places that I think needs additional refactoring, which would probably clear some of this up. The MeetingInfo really needs to be extracted into a separate component connected to the appropriate redux properties. This is definitely making the code more complex than it should be. But that isn't happening right now.

mlippert commented on src/components/Dashboard/MeetingListElement.jsx:
```diff
@@ -54,11 +51,10 @@ class MeetingListElement extends React.Component {
      * @see {@link https://reactjs.org/docs/react-component.html#render|React.Component.render}
      */
     render() {
-        const { uid, meeting, selected, handleMeetingClick } = this.props;
```
I felt it was better to see exactly where props where being used, instead of having a local variable created for them. This de-structuring was originally added as an interim step as I refactored from the previous function form of the component.

mlippert commented on src/components/Dashboard/index.js:
```diff
@@ -26,39 +26,59 @@ import { logger } from 'libs/utils';
-        dispatch(loadMeetingData(uid, meeting._id));
-    },
-});
+const dashboardMapProps = {
```
sorry, I'm not sure what this comment is referring to?

I ended up getting rid of the connectedComponent function because I didn't feel the efficiency of the using the function was worth the mental effort to follow the indirection to understand what was being done. Particularly if that function got moved to the utility lib. Even in the case where there were multiple components being connected (`Metrics/index.js`)

mlippert commented on src/components/Dashboard/MeetingListElement.jsx:
```diff
@@ -52,37 +52,43 @@ class MeetingListElement extends React.Component {
     render() {
         const m = MeetingListElement._formatMeetingTime(new Date(this.props.meeting.startTime));
+        const itemClasses = [ 'timeline-item' ];
+        let markerColor = '#576066'; // not selected
```
I'm pretty sure I did that.

mlippert commented on src/components/MakeMeetingCard/MakeMeetingCardView.jsx:
```diff
@@ -11,51 +27,85 @@ border-radius: 5px;
 
 // linear-gradient(30deg, rgba(84,48,96,1) 20%, rgba(138,106,148,1) 50%, rgba(229,178,245,1) 100%);
-//linear-gradient(66deg, rgba(70,19,85,1) 0%, rgba(138,106,148,1) 69%, rgba(201,155,215,0.9) 100%);
+// linear-gradient(66deg, rgba(70,19,85,1) 0%, rgba(138,106,148,1) 69%, rgba(201,155,215,0.9) 100%);
```
It was there before I started the refactor, and I'm not sure either, so I was/am not ready to delete it.

mlippert commented on src/components/MakeMeetingCard/index.js:
```diff
@@ -1,44 +1,51 @@
-import MakeMeetingCardView from './MakeMeetingCardView';
+import { MakeMeetingCardView } from './MakeMeetingCardView';
 
 const mapStateToProps = state => ({
```
This wasn't a refactor. This was just a fly-by fixing of lint errors, probably because I touched the file for some minor reason like using named colors.

The refactoring was all focused solely on Dashboard files.

mlippert commented on src/components/Dashboard/Metrics/styled.js:
```diff
@@ -0,0 +1,82 @@
+})`
+margin-left: 1rem;
+margin-right: 1rem;
+color: rgb(138,106,148);
```
yep, looks like I missed some.

mlippert commented on src/components/Dashboard/Metrics/TimelineChart.jsx:
```diff
@@ -179,8 +183,29 @@ class TimelineChart extends React.Component {
+     */
+    _onResize(e) {
+        logger.debug('TimelineChart._onResize:', e);
+        this.setState({ chartWidth: document.body.clientWidth / 2 });
```
Just curious, if it wasn't an "expression" but just document.body.clientWidth would you have suggested moving it to a function?
I think you're right that setting the initial value for the chart width, and updating it should be done in the same place, ie it is worth extracting it to a function.

mlippert commented on src/components/Dashboard/Metrics/TimelineChart.jsx:
```diff
@@ -109,24 +130,7 @@ class TimelineChart extends React.Component {
-    }
-
-    /* **************************************************************************
-     * shouldComponentUpdate                                               */ /**
```
No.
Who had those concerns?
The issues from my 1st pass refactor, and when I added the shouldComponentUpdate returning false had to do with how one goes about utilizing d3 within a react component.

Learning more about [shouldComponentUpdate](https://reactjs.org/docs/react-component.html#shouldcomponentupdate), returning false was working for now, but wasn't really the right way to do this anyway. Giving the gantt chart d3 class the ability to redraw and erase and then utilizing that functionality is the correct solution, and is what I finally got around to here (well it's still a poor man's redraw, but it does the job). See the devnote on the file comment for more info.

jaedoucette replied:
The comments preceding the original suggested that this was a performance optimization that had been removed. Brec had a vague recollection of a recent bug related to a component that redrew much too often. We figured we should flag this for a second look.

mlippert commented on src/components/Dashboard/Metrics/gantt.js:
```diff
@@ -28,152 +39,166 @@ const TimeDomainMode = {
+        this._privateProp = {
+            parentSelector: null,
+            svg: null,
+            width: document.body.clientWidth / 2,
```
same expression, but different context, so only tangentially related.

mlippert commented:
OK, I've made changes based on @jaedoucette and @brecriffs 's review comments  and pushed the commit with them.
Check them out and if things look good now, approve the PR please.

jaedoucette commented:
It probably wouldn't have caught my attention as something we should look
at more closely without the compound expression, but Brec and I did discuss
this as being primarily a maintainability issue because of the two
locations.


On Mon, Jun 17, 2019 at 2:22 PM Mike Lippert <notifications@github.com>
wrote:

> *@mlippert* commented on this pull request.
> ------------------------------
>
> In src/components/Dashboard/Metrics/TimelineChart.jsx
> <https://github.com/rifflearning/riff-rtc/pull/121#discussion_r294443322>:
>
> > @@ -179,8 +183,29 @@ class TimelineChart extends React.Component {
>          );
>      }
>
> -    _drawGantt() {
> +    /* **************************************************************************
> +     * _onResize                                                           */ /**
> +     *
> +     * Handle a resize event of the window, which we want to resize the gantt
> +     * chart.
> +     */
> +    _onResize(e) {
> +        logger.debug('TimelineChart._onResize:', e);
> +        this.setState({ chartWidth: document.body.clientWidth / 2 });
>
> Just curious, if it wasn't an "expression" but just
> document.body.clientWidth would you have suggested moving it to a function?
> I think you're right that setting the initial value for the chart width,
> and updating it should be done in the same place, ie it is worth extracting
> it to a function.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/rifflearning/riff-rtc/pull/121?email_source=notifications&email_token=AL4D2X7PSOPTCJ5BLJEISWTP27I6TA5CNFSM4HVYXFL2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3YOQJY#discussion_r294443322>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL4D2XY3ZQS4RJ4WDXIBUCDP27I6TANCNFSM4HVYXFLQ>
> .
>

mlippert commented:
**Release** [1.2.0-dev.9](https://github.com/rifflearning/riff-rtc/releases/tag/1.2.0-dev.9)
